Mixed enhancements

streamlined2 opened the pull request:
Added validation support for parsed entities
Added unit tests for entity/DTO mapper classes
Added IT class ParserIT; JaCoCo collects integration test results
Added unit test for UserServiceImpl class
Renamed 'in memory' storage and repository implementation classes
Removed PII from method 'toString' of entity and DTO classes
Pom.xml gets reformatted
Extracted classes StorageConfiguration, SecurityConfiguration
Single parser component used by service beans to parse entity data

streamlined2 committed:
Added storage, service, repository (DAO) interfaces and classes

1. Added Spring Security dependency to pom.xml for password encoder;
2. SpringcoretaskApplication creates password encoder bean; lists all
available entity rows on application start;
3. Added record classes for entities TraineeDto, TrainerDto,
TrainingDto;
4. Added mapper classes to convert entities to DTOs and vice versa
TraineeMapper, TrainerMapper, TrainingMapper;
5. Added storage classes for each entity TraineeStorage, TrainerStorage,
TrainingStorage;
6. Added repository interfaces TraineeRepository, TrainerRepository,
TrainingRepository and implementation classes DefaultTraineeRepository,
DefaultTrainerRepository, DefaultTrainingRepository;
7. Added service interfaces TraineeService, TrainerService,
TrainingService and implementation classes DefaultTraineeService,
DefaultTrainerService, DefaultTrainingService;
8. Added exception classes EntityAlreadyExistsException,
NoSuchEntityException.

streamlined2 committed:
Storage beans filled by reading and parsing CSV data files

1. Added fasterxml.jackson dependencies to pom.xml;
2. Added CSV data files for trainees, trainers, and trainings;
3. Specified CSV data file names in configuration file
application.properties;
4. Added base CSV data parsing class Parser and derived specialized
classes for every entity TraineeParser, TrainerParser, TrainingParser;
5. Storage beans TrainingStorage, TrainerStorage, TraineeStorage get
initialized by reading and parsing CSV data files;
6. Created base entity interface Entity to provide primary key which
entity classes User, Trainee, Trainer, Training implement;
7. Added no-args constructor to entity classes User, Trainee, Trainer,
Training to support serialization;
8. Created ParseException exception class to signal parse errors;
9. CSVMapper bean created and configured in SpringcoretaskApplication
class.

streamlined2 committed:
Added error handling and logging on service/application level

1. Logging configured in application.properties;
2. Added exception classes EntityCreationException,
EntityDeletionException, EntityUpdateException, EntityQueryException;
3. Subsystem exceptions caught, logged, and rethrown on service layer
level in classes DefaultTraineeService, DefaultTrainerService,
DefaultTrainingService;
4. Exceptions caught, logged, and rethrown on application level.

streamlined2 committed:
Added user name serial number support for Trainee, Trainer entities

1. Added user name serial number support to classes User, Trainee,
Trainer;
2. Added max user name serial number query method getMaxUsernameSerial
to interfaces TrainerRepository, TraineeRepository and implementation
classes DefaultTraineeRepository, DefaultTrainerRepository;
3. Service classes DefaultTraineeService, DefaultTrainerService
increment user name serial number on saving new entity and save previous
value when updating entity.

streamlined2 committed:
Introduced security service interface and implementation class

1. Added security service interface SecurityService and implementation
class DefaultSecurityService which provide new password generation and
encrypting it to password hash;
2. DefaultTraineeService, DefaultTrainerService use security service
component to generate password and obtain its hash;
3. Password type of TraineeService, TrainerService methods converted
from String to char[];
4. Fixed spare spaces in CSV data files.

streamlined2 committed:
Unit tests for methods DefaultTraineeService.deleteById,findById,findAll

streamlined2 committed:
Defined unit tests for every method of DefaultTraineeService class

streamlined2 committed:
Added unit tests for DefaultTrainerService class

streamlined2 committed:
Added unit tests for DefaultTrainingService class

streamlined2 committed:
Added DAO layer unit tests for TraineeRepository class

1. Added DAO layer unit tests for methods of TraineeRepository class;
2. Removed unnecessary PasswordEncoder dependency from TraineeStorage,
TrainerStorage classes;
3. Removed unused constructors from entity classes User, Trainee,
Trainer.

streamlined2 committed:
Added DAO layer unit tests for DefaultTrainerRepository class

1. Added DAO layer unit tests for methods of DefaultTrainerRepository
class;
2. Added no-args constructor and methods 'clear','size' to class
TrainerStorage for testing purposes.

streamlined2 committed:
Fixed auto-wiring error of TrainerStorage bean

1. TrainerStorage class constructor with TrainerParser dependency marked
for auto-wiring.

streamlined2 committed:
Added DAO layer unit tests for class DefaultTrainingRepository

1. Added DAO layer unit tests for methods of class
DefaultTrainingRepository;
2. Fixed errors: assumeTrue replaced with assertTrue in test classes
DefaultTraineeRepositoryTest, DefaultTrainerRepositoryTest;
3. Added no-args constructor and methods 'clear','size' to class
TrainingStorage for testing purposes.

streamlined2 committed:
Fixed code indentation issue, tabs replaced with spaces

streamlined2 committed:
Random generator bean defined, security service parameters externalized

1. Added random generator bean to SpringcoretaskApplication class;
2. Added security service parameters to application.properties
configuration file;
3. Added MissingAlgorithmException class.

streamlined2 committed:
Applied miscellaneous logging style recommendations

1. Logger variable name changed from 'log' to 'LOGGER';
2. Logging level changed from ERROR to DEBUG in DefaultTraineeService,
DefaultTrainerService, DefaultTrainingService classes;
3. Removed exception logging on WARNING level from repository/DAO layer;
4. Method getTrainingKey removed from class Training and replaced with
getPrimaryKey.

streamlined2 committed:
Renamed implementation classes for service,DAO,storage layers

streamlined2 committed:
Added exception throwing declarations to interface definitions

streamlined2 committed:
Added unit test for SecurityServiceImpl class

streamlined2 committed:
Introduced abstract generic base storage class HashMapStorage

1. Added abstract generic base storage class HashMapStorage; concrete
implementation component classes HashMapTraineeStorage,
HashMapTrainerStorage, HashMapTrainingStorage derive from it.

streamlined2 committed:
Removed null passwordHash parameter of Trainer entity in TrainerMapper

streamlined2 committed:
Parser beans created and injected into storage beans via qualifiers

streamlined2 committed:
Storage beans created and injected into repositories via qualifiers

streamlined2 committed:
Added integration tests for TraineeService service

streamlined2 committed:
Integration test class renamed into SpringcoretaskApplicationIT

streamlined2 committed:
User name serial number calculated using both trainee/trainer storage

1. Introduced base service class UserServiceImpl which contains business
logic of finding next user name serial number using TraineeRepository,
TrainerRepository dependencies; service classes TraineeServiceImpl,
TrainerServiceImpl derived from it;
2. Business logic method 'setNextUsernameSerial' moved from entity class
User into service class UserServiceImpl.

streamlined2 committed:
Removed Spring Boot starters and replaced with specific dependencies

streamlined2 committed:
Storage beans initialization accomplished by service beans

streamlined2 committed:
Added dependency for jacoco-maven-plugin

streamlined2 committed:
Single parser component used by service beans to parse entity data

streamlined2 committed:
Extracted classes StorageConfiguration, SecurityConfiguration

streamlined2 committed:
Pom.xml gets reformatted

streamlined2 committed:
Removed PII from method 'toString' of entity and DTO classes

streamlined2 committed:
Renamed 'in memory' storage and repository implementation classes

streamlined2 committed:
Added unit test for UserServiceImpl class

streamlined2 committed:
Added IT class ParserIT; JaCoCo collects integration test results

1. Test configuration parameters application-integration-test.properties
moved to folder src/test/resources;
2. Added integration test class ParserIT; created test data files for
parser integration tests traineestorage*.csv in folder
src/test/resources;
3. Added tests for method 'addAll' to test classes
InMemoryTraineeRepositoryTest, InMemoryTrainerRepositoryTest,
InMemoryTrainingRepositoryTest;
4. Integration test class SpringcoretaskApplicationIT renamed to
TraineeServiceImplIT and moved to folder for service layer;
5. Added dependency for maven-failsafe-plugin to pom.xml; JaCoCo
configured to collect integration test results.

streamlined2 committed:
Added unit tests for entity/DTO mapper classes

streamlined2 committed:
Added method 'isIdenticalTo' to entity classes to compare content

1. Added method 'isIdenticalTo' to interface Entity and implementing
entity classes Trainee, Trainer, Training to compare data they contain;
2. Unit test assertions rewritten using method 'isIdenticalTo'.

Anderwerty commented on src/main/java/com/streamlined/tasks/SecurityConfiguration.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    @Bean
+    Random random() {
```
algorithmName could be method parameter

streamlined2 replied:
Fixed

Anderwerty commented on src/main/java/com/streamlined/tasks/SecurityConfiguration.java:
```diff
@@ -0,0 +1,40 @@
+        try {
+            return SecureRandom.getInstance(algorithmName);
+        } catch (NoSuchAlgorithmException e) {
+            LOGGER.error("Missing random generator algorithm {}", algorithmName, e);
```
redundant logging as if we have exception, if exception we will see all stack trace line 36 in logs or console

streamlined2 replied:
Fixed. Still I'm in doubt if I should avoid logging any caught exception before rethrowing it or this should be done just in this case?

Anderwerty commented on src/main/java/com/streamlined/tasks/entity/User.java:
```diff
@@ -2,84 +2,112 @@
+        this.passwordHash = passwordHash;
+    }
+
+    public boolean userNameStartsWith(String firstName, String lastName) {
```
better to cover such methods with tests

streamlined2 replied:
Done

streamlined2 committed:
Added validation support for parsed entities

1. Added entity validation classes Validator, EntityValidator,
UserValidator, TraineeValidator, TrainerValidator, TrainingValidator,
exception class InvalidEntityDataException;
2. Class Parser checks every parsed entity, puts valid entities into
map, and logs invalid entities;
3. Added unit test in class ParserIT to check for invalid entities;
4. Added unit tests for validator beans TraineeValidatorTest,
TrainerValidatorTest, TrainingValidatorTest;
5. Added unit test UserTest for methods 'userNameStartsWith',
'getUsernameSerial', 'setUsernameSerial' of class User;
6. Configured SLF4J SimpleLogger: specified logging level, appender,
message format;
7. Random generator algorithm name injected as method parameter, removed
excessive logging in SecurityConfiguration class.
